lib/command.js

```diff
@@ -1181,20 +1242,27 @@ Expecting one of '${allowedValues.join("', '")}'`);
   /**
    * Once we have a promise we chain, but call synchronously until then.
    *
-   * @param {Promise|undefined} promise
+   * @param {*} chain
    * @param {Function} fn
-   * @return {Promise|undefined}
+   * @return {*}
    * @api private
    */
 
-  _chainOrCall(promise, fn) {
-    // thenable
-    if (promise && promise.then && typeof promise.then === 'function') {
+  _chainOrCall(chain, fn) {
+    if (isThenable(chain)) {
       // already have a promise, chain callback
-      return promise.then(() => fn());
+      return chain.then(fn);
     }
+
     // callback might return a promise
-    return fn();
+    chain = fn(chain);
+    if (isThenable(chain) && !this._asyncParsing) {
```

Comment:
I had been wondering about whether and how to encourage or enforce parseAsync for async code. Are there benefits to doing the check here rather than in `Command.parse()`?

Reply:
How would you check it in `parse()`? I think you would have to call all user functions with God knows what parameter values, and even that would not be enough because a function can return a thenable in one cases and something else in other ones.

Reply:
> since the promise gets returned for `.parseAsync()` to wrap

That does not apply to promises returned from arg parsers.

Reply:
Well, actually it does indirectly via the calls to `_awaitThenableOptions()` and `_awaitThenableArguments()`, but only when `parseAsync()` has been called. We do not want unnecessary work in `parse()` calls, that is why the methods are no-op when calling `parse()`.